src/engine/mod.rs

```diff
@@ -0,0 +1,657 @@
+    match serde_json::to_string(&json!({ "message": format!("{}", err) })) {
+        Ok(s) => Error::new(ErrorKind::GraphQLError(s), None),
+        Err(e) => {
+            error!("Failed to serialize error object:  {:#?}", e);
+            Error::new(
+                ErrorKind::GraphQLError("INTERNAL SERVER ERROR".to_string()),
+                None,
+            )
+        }
+    }
+}
+
+impl WarpgrapherRequestContext for () {
+    fn new() {}
+}
+
+/// A Warpgrapher GraphQL engine.
+///
+/// The [`Engine`] struct Juniper GraphQL service
+/// on top of it, with an auto-generated set of resolvers that cover basic CRUD
+/// operations, and potentially custom resolvers, on a set of data types and
+/// the relationships between them.  The engine includes handling of back-end
+/// communications with the chosen databse.
+///
+/// [`Engine`]: ./struct.Engine.html
+///
+/// # Examples
+///
+/// ```rust
+/// use warpgrapher::{Engine, Neo4jEndpoint};
+/// use warpgrapher::engine::config::WarpgrapherConfig;
+///
+/// let config = WarpgrapherConfig::default();
+/// let db = Neo4jEndpoint::from_env("DB_URL").unwrap();
+///
+/// let mut engine = Engine::<(), ()>::new(config, db)
+///     .build().unwrap();
+///
+/// ```
+#[derive(Clone)]
+pub struct Engine<GlobalCtx = (), ReqCtx = ()>
+where
+    GlobalCtx: 'static + Clone + Sync + Send + Debug,
+    ReqCtx: 'static + Clone + Sync + Send + Debug + WarpgrapherRequestContext,
+{
+    pub config: WarpgrapherConfig,
+    pub database: String,
+    pub pool: Option<Pool<CypherConnectionManager>>,
+    pub global_ctx: Option<GlobalCtx>,
+    pub resolvers: WarpgrapherResolvers<GlobalCtx, ReqCtx>,
+    pub validators: WarpgrapherValidators,
+    pub extensions: WarpgrapherExtensions<GlobalCtx, ReqCtx>,
+    pub version: Option<String>,
+    root_node: Option<RootRef<GlobalCtx, ReqCtx>>,
+}
+
+impl<GlobalCtx, ReqCtx> Engine<GlobalCtx, ReqCtx>
+where
+    GlobalCtx: 'static + Clone + Sync + Send + Debug,
+    ReqCtx: 'static + Clone + Sync + Send + Debug + WarpgrapherRequestContext,
+{
+    /// Creates a new [`Engine`], with required parameters config and database
+    /// and allows optional parameters to be added using a builder pattern.
+    ///
+    /// [`Engine`]: ./struct.Engine.html
+    /// [`WarpgrapherConfiguration`]: ./config/struct.WarpgrapherConfiguration.html
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use warpgrapher::{Engine, Neo4jEndpoint};
+    /// use warpgrapher::engine::config::WarpgrapherConfig;
+    ///
+    /// let config = WarpgrapherConfig::new(1, Vec::new(), Vec::new());
+    /// let db = Neo4jEndpoint::from_env("DB_URL").unwrap();
+    ///
+    /// let mut engine = Engine::<()>::new(config, db)
+    ///     .build().unwrap();
+    /// ```
+    pub fn new(config: WarpgrapherConfig, database: String) -> Engine<GlobalCtx, ReqCtx> {
+        Engine {
+            config,
+            database,
+            pool: None,
```

Comment:
Per the above, if it doesn't make sense to start an engine without a database pool, this shouldn't ever be None.